### DISCUSSION TITLE
Monitoring strategy

### DISCUSSION THREAD

**Discussion by alexpdp7** [siderolabs/talos#7003] [Q&A] [answered]
I'd like to set up monitoring for Talos clusters, mostly to check for available updates.

I see two strategies to do this:

* [Enable Talos API access from Kubernetes](https://www.talos.dev/v1.3/advanced/talos-api-access-from-k8s/) and create a simple HTTP deployment that exposes an endpoint that compares the installed Talos version to the latest available on GH. It could also do the same for Kubernetes using the Kubernetes API.
* Ship a Talos config along with `talosctl` and `kubectl`, and a script that uses both commands to do the same thing, but outside the cluster.

I think I like best the first option best, but do people have any preference?

Also, is the `os:reader` role designed for this? Should I use `os:admin` instead? I checked the docs and I see no documentation about generating a specific role.

Another option would be to expose a monitoring endpoint. Although, I'd suggest using HTTP instead of gRPC.

**Comment by alexpdp7** [comment]
Ah, I can extract the Talos version from the nodes. Implementing something at https://github.com/alexpdp7/talos-check .

**Reply by alexpdp7** [reply]
... and that repo above now builds the crudest container image possible to do that, has instructions to deploy the image to a cluster, and how to monitor using Nagios' `check_http`.

**Reply by smira** [reply, marked as answer]
Such controller doesn't have to run in the cluster btw, and `os:reader` is enough to query `Version` API.

The way we have this implemented in Omni is the following:

* grab all tags for the `ghcr.io/siderolabs/installer` container
* parse them, and filter the versions using the supported upgrade contracts: https://github.com/siderolabs/talos/blob/9e8603f53b83e326a2529b2c595b619e04f2b85b/pkg/machinery/compatibility/talos_version.go#L41-L77

**Reply by alexpdp7** [reply]
Ah, that is nice! But I'm not handy with Go :)

Well, I polished a bit https://github.com/alexpdp7/talos-check , so it has a CLI tool that generates manifests to deploy it (instead of using `kubectl create` commands. I find it very tedious to compose Kubernetes manifests file. The CLI tool uses a bizarre way to generate the manifests), so I think it can be reasonably usable.

**Reply by alexpdp7** [reply]
(My talos-check repo is now working correctly for this. It's a deployment that runs in cluster, using a cluster role with list/get nodes.)
